p2panda-auth/src/group/mod.rs

```diff
@@ -0,0 +1,821 @@
+/// states from any points in the group history into a new state. This property is what allows us
+/// to process messages in partial- rather than total-order.
+///
+/// Group membership rules are checked when an action is applied to the previous state, read more
+/// in the `group_crdt` module.
+///
+/// This is an implementation of the `AuthGraph` trait which requires a `prepare` and `process`
+/// method. This implementation allows for providing several generic parameters which allows for
+/// integration into different systems and customization of how group change conflicts are
+/// handled.
+///
+/// - Resolver (RS): contains logic for deciding when group state rebuilds are required, and how
+///   concurrent actions are handled.
+/// - Orderer (ORD): the orderer implements an approach to ordering messages, the protocol
+///   requires that all messages are processed in partial-order, but exactly how this is achieved
+///   is not specified.
+/// - Group Store (GS): global store containing states for all known groups.
+impl<ID, OP, RS, ORD, GS> AuthGraph<ID, OP, RS, ORD> for Group<ID, OP, RS, ORD, GS>
+where
+    ID: IdentityHandle + Display,
+    OP: OperationId + Display + Ord,
+    RS: Resolver<GroupState<ID, OP, RS, ORD, GS>, ORD::Message> + Clone + Debug,
+    ORD: Clone + Debug + Ordering<ID, OP, GroupControlMessage<ID, OP>>,
+    GS: Clone + Debug + GroupStore<ID, GroupStateInner<ID, OP, ORD::Message>>,
+{
+    type State = GroupState<ID, OP, RS, ORD, GS>;
+    type Action = GroupControlMessage<ID, OP>;
+    type Error = GroupError<ID, OP, RS, ORD, GS>;
+
+    /// Prepare a next message/operation which should include all meta-data required for ordering
+    /// auth group operations. An ORD implementation needs to guarantee that operations are
+    /// processed after any dependencies they have on the group graph they are part of, as well as
+    /// any sub-groups.
+    ///
+    /// The method `GroupState::heads` and `GroupState::transitive_heads` can be used to retrieve the
+    /// operation ids of these operation dependencies.
+    fn prepare(
+        mut y: Self::State,
+        action: &Self::Action,
+    ) -> Result<(GroupState<ID, OP, RS, ORD, GS>, ORD::Message), GroupError<ID, OP, RS, ORD, GS>>
+    {
+        // Get the next operation from our global orderer. The operation wraps the action we want
+        // to perform, adding ordering and author meta-data.
+        let ordering_y = y.orderer_y.clone();
+        let (ordering_y, operation) = match ORD::next_message(ordering_y, &action) {
+            Ok(operation) => operation,
+            Err(_) => panic!(),
+        };
+
+        // Queue the operation in the orderer.
+        //
+        // Even though we know the operation is ready for processing (ordering dependencies are
+        // met), we need to queue it so that the orderer progresses to the correct state.
+        //
+        // TODO: maybe this should rather happen inside ORD::next_message?
```

Comment:
I agree, we usually don't need to make the orderer aware of our own messages (they are always already processed), or at least we shouldn't need to enforce this already on this level